tests/typecheck/models/test_abstract.yml

```diff
@@ -51,6 +51,7 @@
     Recursive(parent=Recursive(parent=None))
     Concrete(parent=Concrete(parent=None))
   out: |
+    main:4: error: Access to generic instance variables via class is ambiguous
```

Comment:
Do we know why this new error appears in abstract models?

Are there any use cases for accessing `objects` of an abstract model? I can't think of any.

Reply:
I think the mypy plugin is why we only get this for abstract models, but I haven’t looked at the details.

An abstract model does not have an `objects` attribute at all at runtime, so extra errors for that are fine.

Reply:
> Are there any use cases for accessing `objects` of an abstract model? I can't think of any.

I added this test in #1393. Just want to point out that accessing `objects` on an abstract model wasn't part of any use case, but rather an attempt for completeness of coverage in the plugin code

Reply:
I tried this out: actually one of my projects relies on this in several places, and new `django-stubs` now causes errors. Let me open an issue describing the use case and we can discuss it there.

Reply:
Please have a look at issue https://github.com/typeddjango/django-stubs/issues/1653

Reply:
Hm, the error also appears for e.g. `mypy==1.4.1`. Error is emitted by mypy [here](https://github.com/python/mypy/blob/3534bacc4c0d3c4b1983a533e2a36cce43f2ec9d/mypy/checkmember.py#L994) and they have [an example](https://github.com/python/mypy/blob/3534bacc4c0d3c4b1983a533e2a36cce43f2ec9d/mypy/checkmember.py#L980-L985) in a comment close by looking like this:

```python
# Check if original variable type has type variables. For example:
class C(Generic[T]):
    x: T
C.x  # Error, ambiguous access
C[int].x  # Also an error, since C[int] is same as C at runtime
```

I think `C.x` should be the case that we run in to here(?)

The plugin is populating managers based on the runtime, but since abstract models don't have any, it stays as `objects: BaseManager[Self]`. Otherwise `objects` is (normally) overwritten by the plugin, that's why we're only seeing it for abstract models.

https://github.com/typeddjango/django-stubs/blob/85e1f3f0e14cf64bde9c00da5de6cef87cb01b38/mypy_django_plugin/transformers/models.py#L316-L339

So, I would suspect that if you try to annotate any manager to a model, this error would appear. e.g.

```python
from django.db import models
from typing_extensions import Self

class MyModel(models.Model):
    other_manager: models.Manager[Self]

reveal_type(MyModel.other_manager)
```

Is there something up with `typing_extensions.Self` here, or do we need to change something?

Reply:
You can make mypy accept your example using [`ClassVar`](https://docs.python.org/3/library/typing.html#typing.ClassVar) ([PEP 526](https://peps.python.org/pep-0526/)).

```python
class MyModel(models.Model):
    other_manager: ClassVar[models.Manager[Self]]
```

Another way that more accurately reflects what’s happening at runtime might be

```python
ModelT = TypeVar("ModelT", bound=models.Model)

class ManagerDescriptor:
    def __get__(self, instance: None, owner: type[ModelT]) -> models.Manager[ModelT]: ...

class MyModel(models.Model):
    other_manager: ManagerDescriptor = models.Manager()
```

This PR has no effect on your `other_manager`, of course. I didn’t annotate `django.db.models.Model.objects` itself as a `ClassVar` or descriptor because that would seem to break more things than it fixes: every subclass of `Model` that overrides `objects` would then be required to annotate it as a `ClassVar` or descriptor too.

(Maybe it would be better to delete the `objects` annotation from the stub and rely entirely on the plugin for it?)

Reply:
> Another way that more accurately reflects what’s happening at runtime might be
> 
> ```python
> ModelT = TypeVar("ModelT", bound=models.Model)
> 
> class ManagerDescriptor:
>     def __get__(self, instance: None, owner: type[ModelT]) -> models.Manager[ModelT]: ...
> 
> class MyModel(models.Model):
>     other_manager: ManagerDescriptor = models.Manager()
> ```

I don't think this is correctly reflected? The runtime type of `objects` is actually `models.Manager`, so it's rather that `BaseManager.__get__` that should be implemented? You'd never get hold of the descriptor class itself. Anyways, adding that method doesn't change anything either..

This is a side note but I think that should be an `@overload` method with anything other than class access having `-> NoReturn:`

> (Maybe it would be better to delete the `objects` annotation from the stub and rely entirely on the plugin for it?)

Yes, I've been thinking about this too and I think it's something we _should_ do. I did it just recently with `DoesNotExist` and `MultipleObjectsReturned` in #1663. Though we're leaving other type checkers behind by removing `objects: ` from the `.pyi`. Would've been nice if that attribute would stay, even though it's a bit incorrect, mainly for their quality of life situation.

Reply:
Although, perhaps it's fair to say that it should be declared as a `ClassVar`? The plugin can auto generate properly etc. But doing it explicitly on a subclass one would have to declare it a `ClassVar`?

Reply:
I took the liberty to extend the work here and try out changing managers to `ClassVar`s and adjust `objects` to one too. While the plugin will only add `objects` to models where it exists runtime.

It can be found in #1672